### ISSUE TITLE
SFTP preserve always follows symlinks

### ISSUE THREAD

**Issue by eyalgolan1337** [ronf/asyncssh#666]
When copying a symlink with `preserve=True`, the attributes of the target file are copied, instead of the link, even if `follow_symlinks` is `False`.

The attributes are also copied to the target on the destination, which causes copies to fail with `SFTPNoSuchFile` when the symlink is copied before the target.

**Comment by ronf** [comment]
Thanks for the report!

Unfortunately, the SFTP protocol doesn't provide the ability to do the equivalent of FXP_SETSTAT (used for doing a chown, chmod, or utime on a file) without following symlinks. So, even if I tried to make `copy()` aware of this, it would only work when copying data from a remote system to the local system, and not vice-versa. It also wouldn't work when running as a server and getting a copy request from a remote system, and even when copying to the local system it seems that `os.chmod()` not following links is not supported on all OSes. In particular, Linux explicitly forbids this for `os.chmod()`, and Python raises `NotImplementedError` there:

```
>>> os.chmod('file', 0o555, follow_symlinks=False)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
NotImplementedError: chmod: follow_symlinks unavailable on this platform
```

The Linux chmod man page says the following:

> **chmod** never changes the permissions of symbolic links; the **chmod** system
  call cannot change their permissions.  This is not a problem since  the
  permissions  of  symbolic links are never used.  However, for each symbolic
  link listed on the command line, chmod changes the permissions of
  the pointed-to file.  In contrast, **chmod** ignores symbolic links encoun‐
  tered during recursive directory traversals.

So, there's really no way to make this work in all cases with `follow_symlinks=False` and `preserve=True`. I could make it work better when copying from remote to local, and potentially ignore the exception on Linux to avoid the other issue you mentioned about attempting to follow the link and getting an error that the file doesn't exist, but I don't see a good way in SFTP to set attributes on a remote symlink without following the link. I'd probably need to completely skip doing the preserve and not try to set attributes at all when copying symlinks to a remote system.

**Comment by ronf** [comment]
Ah, thanks - I hadn't noticed that OpenSSH added `lsetstat@openssh.com`. That actually makers me feel better about trying to support this. If this extension is added to AsyncSSH, both it and OpenSSH should have no problem fully implementing update of uid, gid, and timestamps on symlinks on all platforms, and even permissions on symlinks on BSD and macOS.

The only time we'd need to skip updating the attributes  is when `follow_symlinks=False` and `preserve=True` are set and the peer doesn't support lsetstat. In other words, we'd be ignoring `preserve=True` in this case, but only on symlinks.

If I get some time this weekend, I'll take a closer look.

**Comment by ronf** [comment]
Client and server support for lsetstat is now available in the "develop" branch in commit 16fb0ac. This also adds support in `SFTPClient` for the "follow_symlinks" argument on stat(), setstat(), chown(), chmod(), and utime() (similar to the `os` module) and adds a new lsetstat() method in `SFTPServer` which SFTP server applications can implement if they need to customize the behavior of this request.

As part of this, the get/put/copy functions have been changed to better handle when follow_symlinks is set to `False`, and to better handle preserving attributes on symlinks in cases where it is supported.

Note: Setting attributes on symlinks is not possible in Windows at the moment, and it is not possible to set permissions on symlinks on Linux (though you can set owner/group and timestamps). In such cases, AsyncSSH will skip setting permissions, with an "info" level log message indicating that attributes were not preserved.

Thanks again for reporting this!

**Comment by ronf** [comment]
This feature is now available in AsyncSSH 2.15.0.
